src/logic/safe/store/actions/createTransaction.ts

```diff
@@ -91,231 +62,86 @@ const getSafeTxGas = async (txProps: RequiredTxProps, safeVersion: string): Prom
+const getTxCreationTxArgs = async (
+  {
+    to,
+    valueInWei,
+    txData = EMPTY_DATA,
+    operation = Operation.CALL,
+    safeTxGas,
+    txNonce,
+    safeAddress,
+  }: CreateTransactionArgs,
+  safeVersion: string,
+  safeInstance: GnosisSafe,
+  from: string,
+): Promise<TxArgs> => {
+  return {
+    baseGas: '0',
+    data: txData,
+    gasPrice: '0',
+    gasToken: ZERO_ADDRESS,
+    nonce: parseInt(txNonce?.toString() || (await getNonce(safeAddress, safeInstance)), 10),
+    operation,
+    refundReceiver: ZERO_ADDRESS,
+    safeInstance,
+    safeTxGas: safeTxGas ?? (await getSafeTxGas(safeAddress, safeVersion, txData, to, valueInWei, operation)),
+    sender: from,
+    sigs: getPreValidatedSignatures(from),
+    to,
+    valueInWei,
   }
+}
 
-  async prepare(dispatch: Dispatch, state: AppReduxState, txProps: RequiredTxProps): Promise<void> {
-    // Wallet connection
-    const ready = await onboardUser()
-    if (!ready) {
-      throw Error('No wallet connection')
-    }
-
-    // Selectors
-    const { account } = providerSelector(state)
-    this.from = account
-
-    this.safeVersion = currentSafeCurrentVersion(state)
-    this.safeInstance = getGnosisSafeInstanceAt(txProps.safeAddress, this.safeVersion)
-
-    // Notifications
-    this.notifications = createTxNotifications(txProps.notifiedTransaction, txProps.origin, dispatch)
-
-    // Use the user-provided none or the recommented nonce
-    this.nonce = txProps.txNonce?.toString() || (await getNonce(txProps.safeAddress, this.safeVersion))
-
-    this.txProps = txProps
-    this.dispatch = dispatch
-  }
+const isImmediateExecution = async (
+  safeInstance: GnosisSafe,
+  nonce: string,
+  state: AppReduxState,
+): Promise<boolean> => {
+  const lastTx = getLastTransaction(state)
+  return canExecuteCreatedTx(safeInstance, nonce, lastTx)
 }
 
 export const createTransaction = (
   props: CreateTransactionArgs,
   confirmCallback?: ConfirmEventHandler,
   errorCallback?: ErrorEventHandler,
 ): CreateTransactionAction => {
-  return async (dispatch: Dispatch, getState: () => AppReduxState): Promise<void> => {
-    const sender = new TxSender()
+  return async (dispatch: Dispatch, getState: () => AppReduxState): Promise<void | string> => {
+    const { origin = null, navigateToDeeplink = true, safeAddress, delayExecution } = props
 
-    // Selectors
     const state = getState()
 
-    // Assign fallback values to certain props
-    const txProps = {
-      ...props,
-      txData: props.txData ?? EMPTY_DATA,
-      operation: props.operation ?? Operation.CALL,
-      navigateToTransactionsTab: props.navigateToTransactionsTab ?? true,
-      origin: props.origin ?? null,
-    }
+    const provider = providerSelector(state)
+    const safeVersion = currentSafeCurrentVersion(state)
+    const safeInstance = getGnosisSafeInstanceAt(safeAddress, safeVersion)
+    const txArgs = await getTxCreationTxArgs(props, safeVersion, safeInstance, provider.account)
```

Comment:
I would suggest reducing the number of async functions here by resolving `safeTxGas` earlier and passing it to `getTxCreationTxArgs`

Reply:
Again, this will be possible after tackling the type [here](https://github.com/gnosis/safe-react/issues/3494).